src/configuration.js

```diff
@@ -0,0 +1,252 @@
+        PROXY_HOSTNAME: 'proxyHostname',
+        PROXY_PASSWORD: 'proxyPassword',
+        PROXY_STATUS_URL: 'proxyStatusUrl',
+        PROXY_PORT: 'proxyPort',
+        MAX_OPENED_STORAGES: 'maxOpenedStorages',
+
+        // not supported, use env vars directly:
+        // MEMORY_MBYTES: 'memoryMbytes',
+        // HEADLESS: 'headless',
+        // XVFB: 'xvfb',
+        // CHROME_EXECUTABLE_PATH: 'chromeExecutablePath',
+    };
+
+    static ENV_MAP_REVERSED = Object.entries(Configuration.ENV_MAP).reduce((obj, [key, value]) => {
+        obj[value] = key;
+        return obj;
+    }, {});
+
+    static BOOLEAN_VARS = ['localStorageEnableWalMode'];
+
+    static INTEGER_VARS = ['proxyPort', 'internalPort', 'memoryMbytes', 'containerPort'];
+
+    static DEFAULTS = {
+        defaultKeyValueStoreId: 'default',
+        defaultDatasetId: 'default',
+        defaultRequestQueueId: 'default',
+        inputKey: 'INPUT',
+        proxyHostname: 'proxy.apify.com',
+        apiBaseUrl: 'https://api.apify.com/v2',
+        proxyStatusUrl: 'http://proxy.apify.com',
+        proxyPort: 8000,
+        containerPort: 4321,
+        containerUrl: 'http://localhost:4321', // Must match `containerPort` above!
+        maxOpenedStorages: 1000,
+        metamorphAfterSleepMillis: 300e3,
+        persistStateIntervalMillis: 60e3, // This value is mentioned in jsdoc in `events.js`, if you update it here, update it there too.
+        localStorageEnableWalMode: true,
+    };
+
+    /**
+     * Creates new `Configuration` instance with provided options. Env vars will have precedence over those.
+     *
+     * @param {Record<string, number | string | boolean>} options
+     */
+    constructor(options = {}) {
+        this.options = options;
+
+        if (!this.get('localStorageDir') && !this.get('token')) {
+            const dir = join(process.cwd(), './apify_storage');
+            this.set('localStorageDir', dir);
+            // eslint-disable-next-line max-len
+            log.warning(`Neither ${ENV_VARS.LOCAL_STORAGE_DIR} nor ${ENV_VARS.TOKEN} environment variable is set, defaulting to ${ENV_VARS.LOCAL_STORAGE_DIR}="${dir}"`);
+        }
+    }
+
+    /**
+     * Returns configured value. First checks the environment variables, then provided configuration,
+     * fallbacks to the `defaultValue` argument if provided, otherwise uses the default value as described
+     * in the above section.
+     *
+     * @param {string} key
+     * @param {string | number | boolean} [defaultValue]
+     * @return {string | number | boolean}
+     */
+    get(key, defaultValue) {
+        // prefer env vars
+        const envKey = Configuration.ENV_MAP_REVERSED[key] ?? key;
+        const envValue = process.env[envKey] ?? process.env[ENV_VARS[envKey] ?? process.env[`APIFY_${envKey}`]];
+
+        if (envValue != null) {
+            return this._cast(key, envValue, true);
+        }
+
+        // check instance level options
+        if (key in this.options) {
```

Comment:
This checks also the prototype chain. Probably nothing that could happen, but it makes me feel awkward.

Reply:
Why should that matter? The `options` will be always a POJO (we can ensure that by making a shallow copy in the ctor if you want).

Sounds like you are affraid of using another build in JS feature? :P 

We can also have a `Map` instance there, if it really makes you feel awkward

Reply:
replaced with `Map`